[mlir][LLVM] handle ArrayAttr for constant array of structs

jeanPerier opened the pull request:
While LLVM IR dialect has a way to represent arbitrary LLVM constant array of structs via an insert chain, it is in practice very expensive for the compilation time as soon as the array is bigger than a couple hundred elements. This is because generating and later folding such insert chain is really not cheap.

This is an issue for flang because it is very easy to generate rather big array of struct constant initializer in Fortran, and unlike C++, dynamic initialization of globals is not a feature of the language. Initializers must be static.

For instance, here are the compile time I measuring for the following program changing N size:

```
! test.F90
module m
  type t
    integer :: i = 42
    real :: x = 1.0
  end type
  type(t) :: some_global(N)
end module
```


```
/usr/bin/time flang -c -DN=1000 test.F90 
0.08user 0.07system 0:00.11elapsed 140%CPU (0avgtext+0avgdata 83968maxresident)k
8inputs+40outputs (13major+4840minor)pagefaults 0swaps

/usr/bin/time  flang -c -DN=10000 test.F90 
1.40user 0.08system 0:01.44elapsed 102%CPU (0avgtext+0avgdata 89088maxresident)k
8inputs+184outputs (13major+8764minor)pagefaults 0swaps

/usr/bin/time  flang -c -DN=100000 test.F90 
137.79user 0.22system 2:18.00elapsed 100%CPU (0avgtext+0avgdata 145540maxresident)k
8inputs+1584outputs (10major+82461minor)pagefaults 0swap
```

In the last case, more than 99.99% of the time is spend folding the insert chain in ModuleTranslation.cpp
With this patch (and updating flang to generate an `ArrayAttr` instead of an insert chain), the last case with 100000 elements takes 0.15s to compile (~1000x compilation speed up :)).

This is not a silver bullet because there are cases where an insert chain will still currently be needed, like when the initial values contain symbol reference, but this is not very common for my use case.

jeanPerier committed:
[mlir][LLVM] handle ArrayAttr for constant array of structs

Dinistro commented on mlir/test/Dialect/LLVMIR/invalid.mlir:
```diff
@@ -1850,3 +1850,25 @@ llvm.func @gep_inbounds_flag_usage(%ptr: !llvm.ptr, %idx: i64) {
+// -----
+
+llvm.mlir.global @x1() : !llvm.array<2x!llvm.struct<(i32, f32)>> {
+  // expected-error@+1{{'llvm.mlir.constant' op array attribute size does not match array type size in dimension 0: 1 vs. 2}}
```
```suggestion
  // expected-error@below{{'llvm.mlir.constant' op array attribute size does not match array type size in dimension 0: 1 vs. 2}}
```
Nit: Same for the other test cases.

Dinistro commented on mlir/test/Dialect/LLVMIR/invalid.mlir:
```diff
@@ -1850,3 +1850,25 @@ llvm.func @gep_inbounds_flag_usage(%ptr: !llvm.ptr, %idx: i64) {
+
+// -----
+
+llvm.mlir.global @x1() : !llvm.array<2x!llvm.struct<(i32, f32)>> {
```
Nit: Maybe use a slightly more descriptive function name?

Dinistro commented on mlir/test/Target/LLVMIR/llvmir.mlir:
```diff
@@ -3000,3 +3000,20 @@ llvm.func internal @i(%arg0: i32) attributes {dso_local} {
+
+// CHECK: @test_array_attr_2 = global [2 x { i32, float }] [{ i32, float } { i32 42, float 1.000000e+00 }, { i32, float } { i32 42, float 1.000000e+00 }]
+llvm.mlir.global @test_array_attr_2() : !llvm.array<2x!llvm.struct<(i32, f32)>> {
+  %0 = llvm.mlir.constant([[42 : i32, 1.000000e+00 : f32],[42 : i32, 1.000000e+00 : f32]]) : !llvm.array<2x!llvm.struct<(i32, f32)>>
```
```suggestion
  %0 = llvm.mlir.constant([[42 : i32, 1.000000e+00 : f32], [42 : i32, 1.000000e+00 : f32]]) : !llvm.array<2 x !llvm.struct<(i32, f32)>>
```
Ultra nit: Using spaces like this should make the tests slightly more readable. Feel free to ignore, though.

gysit reviewed (commented):
There have been multiple gradual additions to constant operation in the past and I wonder if it would be worth to try to come up with some alternative modeling for such large constants. 

Do you need the full flexibility of this PR or are you mainly interested in a solution of arrays of structs? If it is the later, I would favor making the PR more restrictive to only support what we need.

Also note that the doc string of the constant operation well need to be update if we extend the supported element types.

jeanPerier commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3221,15 +3221,50 @@ LogicalResult LLVM::ConstantOp::verify() {
+                 << dim << ": " << arrayAttr.size() << " vs. " << dimSize;
+        if (arrayAttr.size() == 0)
+          break;
+        dimVal = arrayAttr.getValue()[0];
```
> My understanding is we want to support, potentially large, arrays with different element types? 

No, the arrays element all have the same types, but the element type is a struct type which itself contains different element types (which can be any of int, floats, pointers, struct, arrays).


I do not think there is an attribute other than the ArrayAttr that allow representing such array of values.
The issue is that verifying that all the ArrayAttr have the same "layout" (attribute kind, and type/shape if relevant) will be complex and expensive for big arrays, so the choice here is to make a sanity check of one element.

jeanPerier replied:
New verifier helper does verify all array elements, with a set to avoid rechecking the same attributes again and again. I tested on my flang use case, and the extra cost was reasonable (~+1% compilation time on big array constants).

gysit commented:
> By more restrictive, do you mean rejecting the ArrayAttr on an mlir::Constant for an array whose element is not a struct type?

Yes that was my though. I believe for floats / integers etc we use a ElementsAttr or even a SplatElementsAttr to ensure the types are consistent. ArrayAttr should really be our last resort since consistency is hard to verify. 

> I did not do it because although the ModuleTranslation did not support it, the verifier was already accepting those, so I wondered if someone had a use case for that purely in the LLVM MLIR dialect.

I assume this is a gap in the verifier but maybe there are users nevertheless. 

> I am open to alternatives if you see a better path for my goal.

I think highest priority is probably trying to make the verifier more readable.

Second priority would be having something like a SplatLLVMStructElementsAttr / DenseLLVMStructElementsAttr. That would offload a lot of the verification work to these attributes and would make the verifier much nicer. Also a splat attribute would be much more efficient if all elements are initialized to the same value. 

Unfortunately, I am not sure how easy it is to add such attributes that extend the existing built-in splat / dense array attributes. At least, there seem to be some base classes that probably could be extended (such as `ElementsAttrBase`). However, properly supporting nesting for arrays and structs etc is probably quite difficult. So for sure consider this optional.

@Dinistro @matthias-springer do you have an idea if it would be feasible / make sense to strive for something like Splat/Dense LLVMStruct/LLVMArray ElementsAttr?

jeanPerier committed:
support ZeroAttr and UndefAttr translation to llvm::Constant

jeanPerier committed:
address test formatting comments

jeanPerier committed:
refactor and restrict array of struct constant verification

jeanPerier committed:
update llvm.constant doc

jeanPerier commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3221,15 +3221,50 @@ LogicalResult LLVM::ConstantOp::verify() {
+                  dimType = t.getElementType();
+                  return t.getNumElements();
+                })
+                .Default([](auto) -> int64_t { return -1; });
```
Not needed with more restrictive verifier.

jeanPerier commented:
> I think highest priority is probably trying to make the verifier more readable.

I refactored the verifier code to check array of struct attribute in a helper, which allow recursive checking of sub-dimension making things clearer.

I restricted the verifier to reject ArrayAttr for array of simple element types, and removed the handling of vector type from my code since it does not make sense to have vectors of structs.

I also added support and tests for undef/zero attributes appearing in some dimension or struct element field since I do need this in my flang use case.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3142,6 +3142,72 @@ static bool hasScalableVectorType(Type t) {
+  // Forbid usages of ArrayAttr for simple array types that should use
+  // DenseElementsAttr instead. Note that there would be a use case for such
+  // array types when one element value is obtained via a ptr-to-int conversion
+  // from a symbol and cannot be represented in a DenseElementsAttr, but no MLIR
```
I believe the import from LLVM IR will generate operations to model this. I guess importing some large IR file will tell us if the verifier is now too restrictive (@bcardosolopes FYI). But let's give it a shot and then relax the verifier again if needed.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3142,6 +3142,72 @@ static bool hasScalableVectorType(Type t) {
+  // Shallow verification that leaf attributes are appropriate as struct initial
+  // value.
+  size_t numStructElements = structType.getBody().size();
+  for (auto [idx, elementAttr] : llvm::enumerate(arrayAttr))
+    if (elementsVerified.insert(elementAttr).second) {
+      if (isa<LLVM::ZeroAttr, LLVM::UndefAttr>(elementAttr))
+        continue;
+      auto subArrayAttr = dyn_cast<ArrayAttr>(elementAttr);
+      if (!subArrayAttr)
+        return op.emitOpError()
+               << "nested attribute for struct element at index " << idx
+               << " must be a zero, or undef, or array attribute";
+      if (subArrayAttr.size() != numStructElements)
+        return op.emitOpError()
+               << "nested array attribute size for struct element at index "
+               << idx << " must match struct size: " << subArrayAttr.size()
+               << " vs. " << numStructElements;
+    }
```
```suggestion
  for (auto [idx, elementAttr] : llvm::enumerate(arrayAttr)) {
    if (elementsVerified.insert(elementAttr).second) {
      if (isa<LLVM::ZeroAttr, LLVM::UndefAttr>(elementAttr))
        continue;
      auto subArrayAttr = dyn_cast<ArrayAttr>(elementAttr);
      if (!subArrayAttr)
        return op.emitOpError()
               << "nested attribute for struct element at index " << idx
               << " must be a zero, or undef, or array attribute";
      if (subArrayAttr.size() != numStructElements)
        return op.emitOpError()
               << "nested array attribute size for struct element at index "
               << idx << " must match struct size: " << subArrayAttr.size()
               << " vs. " << numStructElements;
    }
  }
```
ultra nit: missing braces

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -3142,6 +3142,72 @@ static bool hasScalableVectorType(Type t) {
+    return op.emitOpError() << "for array with an array attribute must have a "
+                               "struct element type";
+
+  // Shallow verification that leaf attributes are appropriate as struct initial
```
Shallow bcs we are not verifying the members of the first struct we hit. I guess that is an ok compromise.

gysit reviewed (approved):
Thanks for addressing the comments.

LGTM module nits.

I think the verifier is more readable now and the compromise in terms of verification quality seem ok. At a first glance, it also looks like the LLVM IR import should not be able to produce code that does not pass the new verifier (let's see if this statement survives the test of time...).

Let's give other reviewers some time tomorrow to follow-up should there be concerns.

jeanPerier committed:
Apply suggestions from code review

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

jeanPerier committed:
format
